lib/jekyll/commands/serve/servlet.rb

```diff
@@ -29,31 +29,38 @@ def search_file(req, res, basename)
         def search_index_file(req, res)
           # /file/index.html -> /file.html
 
-          # First, let's see if the default implementation can figure it out.
-          # (Check for index files in the res.filename directory)
-          if file = super
-            return file
+          # First, let's see if the our super method can figure it out.
+          # (i.e Check for index.html files in the res.filename directory)
+          file = super
+
+          unless file
+            # Ok, I guess that didn't work, I guess there's no basename/index.html
+            # Let's look for basename.html instead...
+
+            # Keep a backup of res.filename in case we need to revert our changes to it
+            old_filename = res.filename
+
+            # We need to calculate the basename and remove it from the path (res.filename)
+            # so we'll turn res.filename into an array of path elements then pop off the
+            # basename.
+            #
+            # We use a while loop just in case res.filename has trailing slashes.
+            #
+            # Once we have popped off the basename, we can join up what's left and use it
+            # as the new res.filename.
+            path_arr = res.filename.scan(%r!/[^/]*!)
+            while basename = path_arr.pop
+              break unless basename == "/"
+            end
```

Comment:
`hostname == arr.pop until hostname == "/"` doesn't assign the value of `arr.pop` to `hostname`. Also `hostname` will be undefined when evaluating the `until` condition.

Looking at the `set_filename` implementation, (which sets up the `res.filename` and calls `search_index_file`), `res.filename` is fairly sane and unlikely to change.

It looks like the filename shouldn't have any trailing slashes (so `basename = arr.pop` without testing for trailing slashes may work) when we get in `search_index_file`, but I haven't tested.

Still, the `while` method isn't complex or slow and it makes our method more robust and resilient to change.